### PR TITLE
Update metrics.py: Fix SparseCategoricalAccuracy.update_state() doc string

### PR DESCRIPTION
**Pull request by ziruizhuang**
Update metrics.py: Fix SparseCategoricalAccuracy.update_state() doc string

Try fixing issue #49252 

For SparseCategoricalAccuracy, `y_true` should be integer labels and `y_pred` should be probabilities.

```
    m = tf.keras.metrics.SparseCategoricalAccuracy()

    m.update_state([[2], [1]], [[0.1, 0.6, 0.3], [0.05, 0.95, 0]])  # Correct usage,  `y_true` as integer labels and `y_pred` as probabilities.
    print(m.result().numpy())
    # >>> 0.5

    m.update_state([[2], [1]], [[1], [1]])  # Wrong usage, both as integer labels.
    print(m.result().numpy())
    # >>> 0.25

    m.update_state([[0, 1, 0], [0, 1, 0]], [[0.1, 0.6, 0.3], [0.05, 0.95, 0]])  # Wrong usage, both as probabilities.
    print(m.result().numpy())
    # >>> Error
```

### CODE REVIEW

**Comment by ziruizhuang** [comment]
@googlebot I signed it!

File: tensorflow/python/keras/metrics.py [hunk 1]
```diff
@@ -870,6 +870,30 @@ class SparseCategoricalAccuracy(MeanMetricWrapper):
...
+        by the corresponding element in the `sample_weight` vector. If the shape
+        of `sample_weight` is `[batch_size, d0, .. dN-1]` (or can be broadcasted
+        to this shape), then each metric element of `y_pred` is scaled by the
+        corresponding value of `sample_weight`. (Note on `dN-1`: all metric
+        functions reduce by 1 dimension, usually the last axis (-1)).
+
+    Returns:
+      Update op.
+    """
+    super(SparseCategoricalAccuracy, self).update_state(
```
Comment (qlzh727) [inline tensorflow/python/keras/metrics.py:895]: Please explicitly return here.

File: tensorflow/python/keras/metrics.py [hunk 2]
```diff
@@ -870,6 +870,30 @@ class SparseCategoricalAccuracy(MeanMetricWrapper):
...
     super(SparseCategoricalAccuracy, self).__init__(
         sparse_categorical_accuracy, name, dtype=dtype)
+  
+  def update_state(self, y_true, y_pred, sample_weight=None):
+    """Accumulates metric statistics.
+
+    The shapes of `y_true` and `y_pred` are different.
+
+    Args:
+      y_true: Ground truth label values. shape = `[batch_size, 1]`.
```
Comment (qlzh727) [inline tensorflow/python/keras/metrics.py:880]: It doesn't necessary to be [batch_size, 1], it should still have the same rank as the y_pred, just the last dim should be 1.

**Review by qlzh727 — CHANGES_REQUESTED** [review]

File: tensorflow/python/keras/metrics.py (hunk 1 above)
Comment (ziruizhuang) [reply tensorflow/python/keras/metrics.py:895]: I checked all existing codes in metrics.py, and found out that update_state function usually returns an `Update Op`. I suppose it means that the returned operator is used to update the metric state in the metric object.
It is from the update_state function in class Reduce(Metric). Here, it returns the assign_add operator for a member object (self.total or self.count, depending on specific conditions).

If the return value should be marked more explicit, it may means that the annotations for all the update_state functions should be modified as well.

Any suggestions? Appreciated.

File: tensorflow/python/keras/metrics.py (hunk 2 above)
Comment (ziruizhuang) [reply tensorflow/python/keras/metrics.py:880]: How about,
```suggestion
      y_true: Ground truth label values. shape = `[batch_size, d1, .. dN-1]` or shape = `[batch_size, d1, .. dN-1, 1]`.
```

The rank of `y_true` can be short from the rank of `y_pred` by 1. Or, the ranks can be the same.
It is from the sparse_categorical_accuracy(y_true, y_pred) function at
https://github.com/ziruizhuang/tensorflow/blob/d6ee4a941b79a464a9b229f6a22be50baab62d4d/tensorflow/python/keras/metrics.py#L3514-L3517

File: tensorflow/python/keras/metrics.py [hunk 3]
```diff
@@ -877,7 +877,7 @@ def update_state(self, y_true, y_pred, sample_weight=None):
     The shapes of `y_true` and `y_pred` are different.
 
     Args:
-      y_true: Ground truth label values. shape = `[batch_size, 1]`.
+      y_true: Ground truth label values. shape = `[batch_size, d1, .. dN-1]` or shape = `[batch_size, d1, .. dN-1, 1]`.
```
Comment (ziruizhuang) [inline tensorflow/python/keras/metrics.py:880]: From the existing code, 
https://github.com/tensorflow/tensorflow/blob/9410d39d0a9f850c4fb615a1d8b607f1b75d80c3/tensorflow/python/keras/metrics.py#L634-L644
the dimension start from `d0` and ends at `dN`, counting `N+1` dimensions in total. Is this the correct way of annotation dimensions?

If it is `N+1` dimensions by design, to make the dimension annotation consistent with the above code in `metric.py`, here it should be `[batch_size, d0, .. dN-1]` or `[batch_size, d0, .. dN-1, 1]`. 

If it is `N` dimensions, here it should be `[batch_size, d0, .. dN-2]` or `[batch_size, d0, .. dN-2, 1]`, and other occurrences should be corrected as well.

File: tensorflow/python/keras/metrics.py (hunk 1 above)
Comment (ziruizhuang) [reply tensorflow/python/keras/metrics.py:895]: Thank you. I got the point now. 
```suggestion
    return super(SparseCategoricalAccuracy, self).update_state(
```

File: tensorflow/python/keras/metrics.py (hunk 3 above)
Comment (ziruizhuang) [reply tensorflow/python/keras/metrics.py:880]: All right then, in that case, the docstring here should be
```suggestion
      y_true: Ground truth label values. shape = `[batch_size, d0, .. dN-1]` or shape = `[batch_size, d0, .. dN-1, 1]`.
```

File: tensorflow/python/keras/metrics.py [hunk 4]
```diff
@@ -870,6 +870,30 @@ class SparseCategoricalAccuracy(MeanMetricWrapper):
   def __init__(self, name='sparse_categorical_accuracy', dtype=None):
     super(SparseCategoricalAccuracy, self).__init__(
         sparse_categorical_accuracy, name, dtype=dtype)
+  
+  def update_state(self, y_true, y_pred, sample_weight=None):
+    """Accumulates metric statistics.
```
Comment (qlzh727) [inline tensorflow/python/keras/metrics.py:875]: In fact, this docstring update should be applied to some other metric class as well. Can we take this string out as a private const string on the top, and assign it to the __docstring__ field to all the classes that need it?

eg:

```
_UPDATE_STATE_DOCSTRING = """ ... """"

SparseCategoricalAccuracy.update_state.__docstring__ = _UPDATE_STATE_DOCSTRING

```

Comment (ziruizhuang) [reply tensorflow/python/keras/metrics.py:875]: Actually, the docstring for class `SparseCategoricalAccuracy` added in this PR is copied from the base class `MeanMetricWrapper`, with modification on the dimension description. 
https://github.com/tensorflow/tensorflow/blob/8d83bffd4776bbfad12a4e2669e8e30f7a26d3a7/tensorflow/python/keras/metrics.py#L825

The base class `MeanMetricWrapper` docstring:
https://github.com/tensorflow/tensorflow/blob/8d83bffd4776bbfad12a4e2669e8e30f7a26d3a7/tensorflow/python/keras/metrics.py#L628-L644

The child class `SparseCategoricalAccuracy`  docstring:
https://github.com/tensorflow/tensorflow/blob/8d83bffd4776bbfad12a4e2669e8e30f7a26d3a7/tensorflow/python/keras/metrics.py#L874-L890

It is a good idea to take the docstring out to be reused for other `update_state` functions. For example, it seems all Sparse Categorical-based metrics may suffer from this dimension problem and can reuse this docstring. 

On the implementation, taking a doctring to the top and assign it to the `update_state` functions will definitely works. I was wondering if we can use inheritance here as well, adding a pure wrapper, such as

```
class MeanMetricWrapper:
    def update_state(self, y_true, y_pred, sample_weight=None):
        """DocString Version Default
        """
        print('Done MeanMetricWrapper update state')


class SparseCategoricalMetricWrapper(MeanMetricWrapper):
    def update_state(self, y_true, y_pred, sample_weight=None):
        """DocString Version SparseCategorical
        """
        return super(SparseCategoricalMetricWrapper, self).update_state(y_true, y_pred, sample_weight)


class SparseCategoricalAccuracy(SparseCategoricalMetricWrapper):
    pass


if __name__ == '__main__':
    metric = SparseCategoricalAccuracy()
    print(metric.update_state.__doc__)
    metric.update_state(None, None)

>>> 
DocString Version SparseCategorical
        
Done MeanMetricWrapper update state
```

BTW, maybe a little bit off topic, it may be better to take the docstring of the base class out as well. Because the docstring should be highly dependent on the `_fn` member function upon inheritance of the base class `MeanMetricWrapper`. However, it does seem that for now all child metrics of `MeanMetricWrapper`, excluding the sparse categorical ones, require same ranks in dimensions. So practically, we can leave the `MeanMetricWrapper` code as is for the moment. 
https://github.com/tensorflow/tensorflow/blob/8d83bffd4776bbfad12a4e2669e8e30f7a26d3a7/tensorflow/python/keras/metrics.py#L623-L626
https://github.com/tensorflow/tensorflow/blob/8d83bffd4776bbfad12a4e2669e8e30f7a26d3a7/tensorflow/python/keras/metrics.py#L657-L660

Comment (ziruizhuang) [reply tensorflow/python/keras/metrics.py]: Just added a wrapper class, and extended the class to SparseCategoricalAccuracy, SparseTopKCategoricalAccuracy and SparseCategoricalCrossentropy.

File: tensorflow/python/keras/metrics.py [hunk 5]
```diff
@@ -682,6 +682,44 @@ def from_config(cls, config):
     return super(MeanMetricWrapper, cls).from_config(config)
 
 
+class SparseCategoricalMetricWrapper(MeanMetricWrapper):
```
Comment (qlzh727) [inline tensorflow/python/keras/metrics.py:685]: I think having a dedicated class just for having different docstring is bit too overkill, and also increase maintaince overhead. Can we just put the docstring as const above?

Comment (ziruizhuang) [reply tensorflow/python/keras/metrics.py]: Sure can do. Updated the code as instructed.

**Review by qlzh727 — APPROVED** [review]
Thanks for the fix.

**Comment by qlzh727** [comment]
This PR is mirrored to https://github.com/keras-team/keras/commit/4244fba008d686f644b2ea3711e1897edf32432b. Closing it now.
